GitUI/CommandsDialogs/BrowseDialog/GitStatusMonitor.cs

```diff
@@ -394,71 +333,84 @@ private void Update()
             _statusIsUpToDate = true;
             _previousUpdateTime = Environment.TickCount;
 
+            // capture a consistent state in the main thread
+            var statusCancellationToken = _statusCancellationToken;
+            IGitModule module = Module;
+
             ThreadHelper.JoinableTaskFactory.RunAsync(
                 async () =>
                 {
                     try
                     {
-                        await TaskScheduler.Default;
-
-                        _ignoredFilesPending = _ignoredFilesAreStale;
-
-                        // git-status with ignored files when needed only
-                        var cmd = GitCommandHelpers.GetAllChangedFilesCmd(!_ignoredFilesPending, UntrackedFilesMode.Default, noLocks: true);
-                        var output = Module.RunGitCmd(cmd);
-                        var changedFiles = GitCommandHelpers.GetStatusChangedFilesFromString(Module, output);
-
-                        await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                        var changedFiles = await GetChangedFilesAsync().ConfigureAwait(false);
+                        await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(statusCancellationToken);
 
                         UpdatedStatusReceived(changedFiles);
                     }
-                    catch
+                    finally
                     {
-                        await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-
                         _commandIsRunning = false;
-                        CurrentStatus = GitStatusMonitorState.Stopped;
+
+                        // Schedule update every 5 min, even if we don't know that anything changed
+                        CalculateNextUpdateTime(PeriodicUpdateInterval);
                     }
                 })
                 .FileAndForget();
 
-            // Schedule update every 5 min, even if we don't know that anything changed
-            CalculateNextUpdateTime(MaxUpdatePeriod);
-
             return;
 
-            void UpdatedStatusReceived(IReadOnlyList<GitItemStatus> changedFiles)
+            async Task<IEnumerable<GitItemStatus>> GetChangedFilesAsync()
```

Comment:
Can we extract it as `ChangedFilesProvider`?

```
public interface IChangedFilesProvider
{
    Task<IEnumerable<GitItemStatus>> GetAsync();
}

public sealed class ChangedFilesProvider : IChangedFilesProvider
{
    public ChangedFilesProvider(Func<IGitModule> getModule) { ... }

    public async Task<IEnumerable<GitItemStatus>> GetAsync()
    {
        var module = GetModule();
        // ...
    }

    [NotNull]
    private IGitModule GetModule()
    {
        var module = _getModule();

        if (module == null)
        {
            throw new ArgumentException($"Require a valid instance of {nameof(IGitModule)}");
        }

        return module;
    }
}
```

The call site would probably look something like (pseudo code, it's past midnight here):

```
    try
    {
        await TaskScheduler.Default;
        var changedFiles = await _changedFilesProvider.GetAsync().ConfigureAwait(false);
        
        await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(statusCancellationToken);
        UpdatedStatusReceived(changedFiles);
    }
    catch
    {
        await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(statusCancellationToken);
        CurrentStatus = GitStatusMonitorState.Stopped;
        throw;
    }
    finally
    {
        ...
    }
```

This way we can inject it and test it (you knew I'd say this, didn't you? 😄).

Thoughts?

Reply:
I know this was coming, should have written something about it....
You know my answer too, I have considered being more polite but then it would not be me so my standard answer is: I see no benefit in having this tested. 
The time for a test is better spent on other improvements in my opinion.

As usual, I will think about it but probably come to the same conclusion in some time.
The benefit of a test I see is that the implementation can be replaced without changing the behavior, but if we change the implementation of this small part it is to change the behavior.